### ISSUE TITLE
Cannot get recursion example to work

### ISSUE THREAD

**Issue by cesartl** [arrow-kt/arrow#1240]
I'm trying to use run the recursion scheme example locally but I'm having a few problems.

First I can't get the `@instance` annotation to work. I've added `kapt` in maven and it works fine for `@higherkind`.

I got around it by making `IntListPatternFunctor` an object:

```kotlin
object IntListPatternFunctor : Functor<ForIntListPattern> {
    override fun <A, B> IntListPatternOf<A>.map(f: (A) -> B): IntListPatternOf<B> {
        val lp = fix()
        return when (lp) {
            NilPattern -> NilPattern
            is ConsPattern -> ConsPattern(lp.head, f(lp.tail))
        }
    }
}
```

However I get stuck when trying to define the `multiply` function:

```kotlin
fun multiply(list: IntFixList): Int = Fix.recursive().run {
    IntListPatternFunctor.cata(list, multiply)
}
```

The compiler cannot find `recursive` in `Fix` and cannot find `cata` in the `IntListPattern` functor. However I can see `IntListPatternFunctor.hylo` function.

Is there something I am missing?

**Comment by raulraja** [comment]
@cesartl Use Arrow `0.8.2` and `@extension` instead of `@instance`. The docs need to be updated. Let's keep this open and turn it into a doc update ticket. Also those docs probably need to be under `kotlin:ank` blocks in our docs so they are type checked and they don't fall behind.

**Comment by raulraja** [comment]
remove `()` from `extension. should work.

**Comment by raulraja** [comment]
Extensions need to be placed in different modules if you use `@higherkind` where you define `IntListPattern`. otherwise copy the higherkind generated code and don't use `@higherkind` and you can have all in one module.

**Comment by cesartl** [comment]
ok thanks that part worked.

However, I still can't see `cata` and `ana` in the functor instance. Only `hylo` is visible.

Additionally, I can't resolve `Fix.recursive()`

**Comment by cesartl** [comment]
Thanks for the help I finally managed to get it to work. However, as an arrow novice, it was hard to follow; I had to chase bits of information on different pages.
Would you be interested in a PR to make this a bit easier to pick up for an arrow novice?

**Comment by cesartl** [comment]
I'm offering to do it or even collaborate with you if you want. I want to build a talk on how to do recursion schemes with arrow so I'll be going through some simple examples.

Having the perspective of someone new to arrow might help you make it noob friendly :D

**Comment by AdrianRaFo** [comment]
feel free to do a PR, if not I'll mention you on mine to review :wink:

**Comment by nomisRev** [comment]
> I want to build a talk on how to do recursion schemes with arrow so I'll be going through some simple examples.

That is awesome! Looking forward

**Comment by nomisRev** [comment]
Sorry about that... I was on my phone 😅

**Comment by cesartl** [comment]
Could someone clarify something?

What is the use of the run block here:
```kotlin
return Fix.recursive().run{
        IntListPattern.functor().cata(list, multiply)
    }
```

I've noticed it compiles and works by doing just
```kotlin
 return IntListPattern.functor().cata(list, multiply)
```
but it requires an extra import.

Is the run block just to avoid the import?

**Comment by AdrianRaFo** [comment]
they do the same function here the `.run()` receives a `Recursive` extension function which in this case is the same as providing a recursive "context" that works as the import does

**Comment by nomisRev** [comment]
@cesartl the run block can be useful if you want to enable all syntax and don't import anything for a certain function or larger piece of code.

Similar to how you can enable direct syntax to with `with`, which can also be used instead of `run`.

I.e.
```kotlin
with(IO.monad()) {
  just(1)
    .map { it + 1}
   .tupleRight(3) //IO(Tuple2(2, 3))
} 
```

This only requires a import for monad.

**Comment by cesartl** [comment]
thanks I'm starting to get it slowly. As another example I needed a monoid for `Eval` so I did:

```kotlin
fun <A> evalMonoid(m: Monoid<A>): Monoid<Eval<A>> =
        object : Monoid<Eval<A>> {
            override fun empty(): Eval<A> = Eval.now(m.empty())

            override fun Eval<A>.combine(b: Eval<A>): Eval<A> {
                return this.map2(b) { (x, y) ->
                    m.run {
                        x.combine(y)
                    }
                }
            }

        }
```

My understanding is that in the block of `m.run` the extension function `combine` get added to the context of `A`. that's actually a neat way of doing it. It seems to provide more control than the `implicit` hell of scala 😃

**Comment by AdrianRaFo** [comment]
I guess this can be closed. Feel free to open a new issue or ask on the gitter/slack channels for more questions
